### PR TITLE
[sparsity] Change API to take FQNs as configuration

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65295
* __->__ #65296
* #65294
* #65293
* #65292

The original API described in the https://github.com/pytorch/pytorch/issues/59835
assumed that the per-layer configuration would take a module/layer
reference. However, a more useful approach is to refer to the layers
by their fully qualified names (FQN). That allows us to store the
configuration in a file without serializing the models.

We define a layer's FQN as it's "path" within a model. For example,
if one can refer to a model using `model.layer0.sublayerX`, the FQN
of the sublayerX is `'layer0.sublayerX'`.

Test Plan:
```
python test/test_ao_sparsity.py -- TestBaseSparsifier
buck test @mode/opt //caffe2:test -- TestBaseSparsifier
```

Differential Revision: [D31186830](https://our.internmc.facebook.com/intern/diff/D31186830)

### CODE REVIEW

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186830).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186830).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186830).

File: torch/ao/sparsity/sparsifier/base_sparsifier.py [hunk 1]
```diff
@@ -142,16 +142,34 @@ def prepare(self, model, config):
                     else:
                         stack.append(child)
 
+        # TODO: Remove the configuration by reference ('module')
```
Comment (vkuzo) [inline torch/ao/sparsity/sparsifier/base_sparsifier.py:145]: would be good to do this soon, before users start depending on this

**Review by vkuzo — APPROVED** [review]

File: torch/ao/sparsity/sparsifier/base_sparsifier.py (hunk 1 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/base_sparsifier.py:145]: I agree

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186830).

**Pull request closed without merging**